pv211_utils/systems/tfidf.py

```diff
@@ -8,7 +8,7 @@
 
```

Comment:
@MarekToma: Like with [`datasets.py`][1], use the classes from this module in the following Jupyter notebooks and rerun them to see if they are working:
- `notebooks/cranfield.ipynb`
- `notebooks/trec.ipynb`
- `notebooks/arqmath.ipynb`

For `notebooks/trec.ipynb` and `notebooks/arqmath.ipynb`, the free runtime provided by Google Colab will be insufficient. For these, ise [the PV211 JupyterHub][1] instead.

 [1]: https://github.com/MIR-MU/pv211-utils/pull/8/files/e5d5d7da1e36682fdcb11ddd1efafc727fc97109#r1124185893

Reply:
Systems are @VojtechKalivoda responsibility.

Reply:
@VojtechKalivoda, you can commit directly into the `developer` branch to update the pull request. If you don't get to it by Monday, I will merge the PR without this fix and open a ticket with you as the assignee, so that we can fix this in the future. Is that ok with you?

Reply:
@VojtechKalivoda: https://github.com/MIR-MU/pv211-utils/pull/8#discussion_r1133940113 is not entirely correct. We want to only use `TfidfSystem` in `notebooks/cranfield.ipynb`. For `trec.ipynb`, `arqmath.ipynb`, and `notebooks/beir_cqadupstack.ipynb`, we want `BM25Plus` instead. As with the math preprocessing, this is high-priority, since it affects the first term project and should be ready by Monday.

Reply:
The `cranfield`, `trec`, and `beir_cqadupstack` notebooks have been updated to use the new `systems` module. However, there is a problem with ARQMath evaluation - the new evaluation module has really high memory consumption when using multiple workers (it looks like either more data than needed are copied for all threads or/and some weird data leakage). As a result, we are now able to evaluate ARQMath with only one worker, which is way too slow (25-30 minutes on iirhub). This behavior can be mitigated by specifying `maxtasksperchild`, and `chunksize` of Pool, resulting in the terminating and restarting processes after a number of tasks + chunking the data. The problem with this is that optimal values for these parameters are highly dependent on a used dataset, number of workers, etc.

So, in the end, we will either require students to specify all three params (`num_workers`, `maxtasksperchild`, `chunksize`), rewrite the multiprocessing part in the evaluation, or find another solution.

Any preferred solution/ideas? @Witiko @MarekToma

Reply:
I will try to make it more memory efficient.

Reply:
I updated the code to make it hopefully more memory efficient / maybe avoid data leakage. I tested it on num_processes=4 on jupyterhub and both worked and finished in a couple of minutes.